pinot-core/src/main/java/org/apache/pinot/core/transport/server/routing/stats/ServerRoutingStatsManager.java

```diff
@@ -0,0 +1,373 @@
+      Integer numInFlightRequests = stats.getNumInFlightRequests();
+      Double numInFlightRequestsEMA = stats.getInFlightRequestsEMA();
+      Double latencyEMA = stats.getLatencyEMA();
+      Double score = stats.computeHybridScore();
+      stats.getServerReadLock().unlock();
+
+      stringBuilder.append(";").append(server).append("=").append(numInFlightRequests.toString()).append(",")
+          .append(numInFlightRequestsEMA.toString()).append(",").append(latencyEMA.toString()).append(",")
+          .append(score.toString());
+    }
+
+    return stringBuilder.toString();
+  }
+
+  /*
+   * ===================================================================================================================
+   * The helper functions required by various AdaptiveSelectors are defined below.
+   *
+   * 1. NumInFlightReqSelector - fetchNumInFlightRequestsForAllServers(), fetchNumInFlightRequestsForServer()
+   * 2. LatencySelector - fetchEMALatencyForAllServers(), fetchEMALatencyForServer()
+   * 3. HybridSelector - fetchScoreForAllServers(), fetchScoreForServer()
+   *
+   * We avoid returning all the stats to each selector to keep the critical section (under locks) as small as
+   * possible). ServerRoutingStatsManager does not sort the servers in any particular order while accumulating stats
+   * as it is not aware of what sorting strategy to use. This logic is contained in the various
+   * AdaptiveServerSelectors.
+   *
+   * TODO: Explore if reads to the _serverQueryStatsMap can be done without locking.
+   * ===================================================================================================================
+   */
+
+  /**
+   * Returns a list containing each server and the corresponding number of in-flight requests active on the server.
+   */
+  public List<Pair<String, Integer>> fetchNumInFlightRequestsForAllServers() {
+    List<Pair<String, Integer>> response = new ArrayList<>();
+    if (!_isEnabled) {
+      return response;
+    }
+
+    for (Map.Entry<String, ServerRoutingStatsEntry> entry : _serverQueryStatsMap.entrySet()) {
+      String server = entry.getKey();
+      Preconditions.checkState(entry.getValue() != null, "Server stats is null");
+      ServerRoutingStatsEntry stats = entry.getValue();
+
+      stats.getServerReadLock().lock();
+      int numInFlightRequests = stats.getNumInFlightRequests();
+      stats.getServerReadLock().unlock();
+
+      response.add(new ImmutablePair<>(server, numInFlightRequests));
+    }
+
+    return response;
+  }
+
+  /**
+   * Same as above but returns the number of inflight requests for the input server.
+   */
+  public Integer fetchNumInFlightRequestsForServer(String server) {
+    if (!_isEnabled) {
+      return null;
+    }
+
+    ServerRoutingStatsEntry stats = _serverQueryStatsMap.get(server);
+    if (stats == null) {
+      return null;
+    }
+
+    try {
+      stats.getServerReadLock().lock();
+      return stats.getNumInFlightRequests();
```

Comment:
same here

Reply:
Ack. Will be addressed by TODO in subsequent PR.